core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1283,8 +1287,20 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     * res0: Unit = ()
     * }}}
     */
-  def foreach[F2[x] >: F[x]](f: O => F2[Unit]): Stream[F2, Nothing] =
-    flatMap(o => Stream.exec(f(o)))
+  def foreach[F2[x] >: F[x]](f: O => F2[Unit]): Stream[F2, Nothing] = {
+    def goChunk(ch: Chunk[O], ix: Int, s: Int): Pull[F2, Nothing, Unit] =
+      if (ix >= s) Pull.done else Pull.eval(f(ch(ix))) >> goChunk(ch, ix + 1, s)
```

Comment:
Looking at the Pull compile code, we do not make a call to `interruptGuard` when interpreting a `Pull.Eval` node, so it may be there is no difference between running this or traversing the chunk (as done below).